### PR TITLE
Respect error codes from extensions instead of recomputing them

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
The shoot reconciliation/deletion flow does now respect/take over the error codes presented in the extensions' `.status.lastError.codes[]` field instead of unnecessarily recomputing them.
This will help to split up the (partly provider-specific) error code matchers to the respective extension repositories.

**Special notes for your reviewer**:
The `errorWithCode` was replaced by a new `errorWithCodes` structure that allows having multiple error codes instead of only one. Consequently, the `DetermineError` function may return multiple error codes if applicable. 

**Release note**:

```noteworthy developer
Gardener will now respect the error codes presented in the extension CRDs' `.status.lastError.codes[]` field instead of recomputing them.
```

### CODE REVIEW

**Comment by danielfoehrKn** [comment]
Rebase needed :)

**Review by danielfoehrKn — DISMISSED** [review]

**Comment by rfranzke** [comment]
/kind/enhancement

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 1]
```diff
@@ -65,47 +68,59 @@ func DetermineError(err error, message string) error {
...
+
+	// eliminate duplicates
+	var out []gardencorev1beta1.ErrorCode
+	for _, c := range codes.List() {
+		out = append(out, gardencorev1beta1.ErrorCode(c))
+	}
+	return out
 }
 
 // Coder is an error that may produce an ErrorCode visible to the outside.
```
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]:
```suggestion
// Coder is an error that may produce ErrorCodes visible to the outside.
```

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 2]
```diff
@@ -25,23 +25,26 @@ import (
...
-	code    gardencorev1beta1.ErrorCode
+type errorWithCodes struct {
 	message string
+	codes   []gardencorev1beta1.ErrorCode
 }
 
-// NewErrorWithCode creates a new error that additionally exposes the given code via the Coder interface.
-func NewErrorWithCode(code gardencorev1beta1.ErrorCode, message string) error {
-	return &errorWithCode{code, message}
+// NewErrorWithCodes creates a new error that additionally exposes the given code via the Coder interface.
```
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]:
```suggestion
// NewErrorWithCodes creates a new error that additionally exposes the given codes via the Coder interface.
```

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 3]
```diff
@@ -65,47 +68,59 @@ func DetermineError(err error, message string) error {
...
+		codes.Insert(string(gardencorev1beta1.ErrorInfraDependencies))
+	}
+	if resourcesDepletedRegexp.MatchString(message) {
+		codes.Insert(string(gardencorev1beta1.ErrorInfraResourcesDepleted))
+	}
+	if configurationProblemRegexp.MatchString(message) {
+		codes.Insert(string(gardencorev1beta1.ErrorConfigurationProblem))
+	}
+
+	// eliminate duplicates
```
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]: What do you mean by that? Duplicates are already eliminated by using a `set`, right?

File: pkg/apis/core/v1beta1/helper/errors.go [hunk 4]
```diff
@@ -65,47 +68,59 @@ func DetermineError(err error, message string) error {
...
-	var coder Coder
+func determineErrorCodes(err error) []gardencorev1beta1.ErrorCode {
+	var (
+		coder   Coder
+		message = err.Error()
+		codes   = sets.NewString()
+	)
 
-	// first try to re-use code from error
+	// try to re-use code from error
```
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]:
```suggestion
	// try to re-use codes from error
```

**Review by timebertt — COMMENTED** [review]
nits only :)

**Review by timebertt — APPROVED** [review]
Thanks, /lgtm

**Pull request merged**